### PR TITLE
GHCR: Add GitHubPackages.sanitize_root_url

### PR DESCRIPTION
**Pull request by sjackman**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Support `root_url "docker://registry/org/repo"` by translating
    docker://registry/org/repo
to
    https://registry/v2/org/repo

### CODE REVIEW

File: Library/Homebrew/github_packages.rb [hunk 1]
```diff
@@ -101,6 +101,18 @@ def self.root_url(org, repo, prefix = URL_PREFIX)
     "#{prefix}#{org}/#{repo_without_prefix(repo)}"
   end
 
+  def self.sanitize_root_url(url)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/github_packages.rb:104]: I’d just make this part of root_url unless there’s cases where it shouldn’t be used (in which case: what are they)?

File: Library/Homebrew/github_packages.rb [hunk 2]
```diff
@@ -101,6 +101,18 @@ def self.root_url(org, repo, prefix = URL_PREFIX)
...
+    if url.to_s.start_with? "docker://"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/github_packages.rb:105]: May as well do the match here (and same below) so you don’t end up with the prefix being present but the match failing.

File: Library/Homebrew/software_spec.rb [hunk 3]
```diff
@@ -444,18 +444,11 @@ def prefix=(prefix)
 
   def root_url(var = nil, specs = {})
     if var.nil?
-      @root_url ||= if Homebrew::EnvConfig.bottle_domain.match?(GitHubPackages::URL_REGEX)
-        GitHubPackages.root_url(tap.user, tap.repo).to_s
-      else
-        "#{Homebrew::EnvConfig.bottle_domain}/#{Utils::Bottles::Bintray.repository(tap)}"
-      end
+      @root_url ||= GitHubPackages.sanitize_root_url(
+        "#{Homebrew::EnvConfig.bottle_domain}/#{Utils::Bottles::Bintray.repository(tap)}",
```
Comment (MikeMcQuaid) [inline Library/Homebrew/software_spec.rb:448]: I’d rather we didn’t reference bintray here. It’s just going to make it harder to remove that code later.

File: Library/Homebrew/github_packages.rb [hunk 4]
```diff
@@ -101,6 +101,18 @@ def self.root_url(org, repo, prefix = URL_PREFIX)
...
+      _, registry, org, repo = *url.match(%r{docker://([\w.-]+)/([\w-]+)/([\w-]+)})
```
Comment (MikeMcQuaid) [inline Library/Homebrew/github_packages.rb:106]:
```suggestion
    if (matches = url.to_s.match(%r{docker://([\w.-]+)/([\w-]+)/([\w-]+)}))
      _, registry, org, repo = *matches.to_a
```

File: Library/Homebrew/github_packages.rb [hunk 5]
```diff
@@ -101,6 +101,18 @@ def self.root_url(org, repo, prefix = URL_PREFIX)
...
+      GitHubPackages.root_url(org, repo, "https://#{registry}/v2/")
+    elsif url.to_s.start_with? URL_PREFIX
+      _, org, repo = *url.match(URL_REGEX)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/github_packages.rb:109]:
```suggestion
    elsif (matches = url.to_s.match(URL_REGEX))
      _, org, repo = *matches.to_a
```

**Comment by sjackman** [comment]
Superseded by PR #11071 `Fix Linuxbrew URL handling`. Thanks, Mike!

**Pull request closed without merging**